### ISSUE TITLE
Add custom type builders in buildASTSchema

### ISSUE THREAD

**Issue by alexandrebodin** [graphql/graphql-js#1086]
Hello everyone,

After implementing my own schema modifier based on the astNodes of the schema I found that really ineffective not to be able to do it while building the schema. 

A lot of tools do this kind of things (modifying the resolvers etc...) on there own and end up reimplementing in some way the func `buildASTSchema`

I would love to see something like a visitor like option with mappings to build function eg: makeTypDef / makeFieldDef...
and expose some internal helper.

As a conversation starter I made an example of what it could look like
```js
const {
    GraphQLObjectType,
    GraphQLScalarType,
    buildSchema,
    schemaBuilder,
} = require('graphql');

const src = `
type Query {
  hello(id: ID): Hello
}

type Hello {
  id: ID @globalID(prefix: "event")
}
`;


buildSchema(src, {
    builders: {
        makeTypDef(def) {
            const typeName = def.name.value;

            // do your own building logic eg:
            const fields = schemaBuilder.makeFieldDefMap(def);

            Object.keys(fields).forEach(fieldName => {
                const field = fields[fieldName];
                const directives = field.astNode.directives;
                if (directives.some(directive => directive.name.value === 'globalId')) {
                    const globalDirective = directives.find(d => d.name.value === 'globalId');
                    const prefix = globalDirective.arguments.find(arg => arg.name.value === 'prefix');
                    field.resolve = ({ id }) => prefix.value.value + '|' + id;
                }
            });

            const type = new GraphQLObjectType({
              name: typeName,
              description: schemaBuilder.getDescription(def),
              fields: () => fields,
              interfaces: () => schemaBuilder.makeImplementedInterfaces(def),
              astNode: def,
            });
        }
    }
});
```

Standardizing this modifier behavior could be a really good thing to standardize the buildSchema process / schema stitching etc in the community and avoid having multiple implementations of the same mechanism everywhere.

Maybe this belongs in another library though, I'd love to discuss that too.

Thanks !

**Comment by IvanGoncharov** [comment]
I partly implemented this in #1000 by extracting type building code into `ASTDefinitionBuilder` class. It would be great if `buildSchema` function will have an argument to supply derived class to be used during schema building.

**Comment by leebyron** [comment]
Interesting idea! Perhaps you could build atop #1000 or double down on the "visitor" idea by allowing for schema transforms to be written in this fashion.

I think a standalone library would be reasonable, though if the utility is small and highly generalized then it could live here as well.
